# 분산분석 (ANalysis Of VAriance)

## 서론

* 12장에서 두 모집단의 평균을 비교하는 방법을 학습 (T검정을 주로 이용)
* 이번장에서는 여러 모집단에서의 관측자료(평균 등)를 효과적으로 분석하고 해석하게 해 주는 방법에 대해 학습
* 여러 종류의 건전지 중 어느 건전지의 수명이 오래가는지... 여러 종류의 볍씨중 어느 볍씨가 수확량이 좋은지 등...
* 이번장에서는 여러집단의 모평균을 비교하기 위하여 전체 변동을 두가지 변동요인으로 나누어(모집단간 변동, 모집단 내 변동) 분석하는 방법에 대해 학습하며
* 이를 일원배치 분산분석법 이라고 함
    * 종속변수가 1개이며, 독립변수(요인) 가 1개인 경우 일원분산분석(Oneway ANOVA)
    * 종속변수가 1개이며, 독립변수(요인) 가 2개인 경우 이원분산분석(Twoway ANOVA)
    * 종속변수가 1개이며, 독립변수(요인) 가 3개 이상인 경우 다원분산분석(Multiway ANOVA)
    * 종속변수가 2개 이상인 경우 다변량분산분석법(MANOVA)
![분산분석](./images/s14-1.png)

* 분산분석을 하기 위한 조건
    * 독립성 : 모집단간의 오차는 서로 독립이어야 함
    * 정규성 : 모집단의 분포가 정규분포여야 함 (평균을 비교하므로 만족?)
    * 등분산성 : 모집단 간의 분산이 동일해야 함

* 알아둘 것
    * SStr (treatment sum of square) 처리제곱합
    * SSE (error sum of squares) 오차제곱합
    * SST (total sum of squares) 총제곱합
    * 제곱합의 자유도 = 제곱을 해서 더해지는 항의 수 - 각 항들에 의해 만족되는 선형 제약조건의 수
    * 평균제곱 = 제곱합 / 제곱합의자유도
    * 평균처리제곱(MStr) = SStr / k-1 (k 는 처리 총수)
    * 평균오차제곱(MSE) = SSE / n-k (n 는 총 관측치개수)
    * F 분포 = MStr / MSE
    * z-검정은 $\sigma^2$을 알 때 μ를 구하는 것이고, t-검정은 σ2을 모를 때 μ를 구하는 것
    * 카이-제곱 검정은 $\sigma^2$을 구하는 것이고, F-검정은 $\sigma_{1}^2 / \sigma_{2}^2$ 을 구할 때 사용
    
    
## 일원배치 분산분석법    

* 예제 1
    * 안경표면의 손상방지를 위한 코팅처리 A, B, C, D 에 대한 효과 분석 (아래표는 마모도를 측정한 자료)
    * 관측값 = (총평균) + (처리에 의한 편차:모집단 간) + (잔차:모집단 내)
    * $y_{ij}$ = $\bar{y}$ + ($\bar{y_{i}}$ - $\bar{y}$) + ($y_{ij}$ - $\bar{y_{i}}$)
    * 만일 4가지 코팅에서 얻어진 평균 마모도에 차이가 없다면 처리에 의한 편차는 거의 0에 가까워야 하며,
    * 이를 이용하여 처리효과에 대한 변동을 측정하는 양으로 이 값들의 제곱합을 사용하며, 처리제곱합(treatment sum of squares, SStr) 이라 함
    * 위에서 잔차의 경우 관측값과 관측값이 속한 집단 내에서의 평균간의 차이로 랜덤오차라고도 하며
    * 이에 대한 제곱합을 오차제곱합(error sum of squares, SSE) 이라 함
    * 관측치와 전체모든 관측치의 평균에 대한 편차( ($y_{ij}$ - $\bar{y}$) 의 제곱을 총제곱합(total sum of squares, SST) 라고 하며,
    * SST = SStr + SSE 의 공식을 만족함
    * 책 보면서 진행

    * 자유도
        * 제곱합의 자유도 = 제곱을 해서 더해지는 항의 수 - 각 항들에 의해 만족되는 선형 제약조건의 수
    * 총제곱합의 자유도 = 처리제곱합의 자유도 + 오차제곱합의 자유도
    * 제곱합과 자유도를 표의 형태로 정리한 표를 분산분석표(ANOVA table) 라 함
    
    * 관측값을 풀어서 설명한 후 제곱을 하여 제곱합의 형태로 변형

* 제곱합의 간편 계산식
    * $T_{i}$ = $\sum_{j=1}^{n_{i}} y_{ij}$ = 처리 i 에 서의 모든 관측값의 합계
    * $T$ = $\sum_{i=1}^k$ $\sum_{j=1}^{n_{i}} y_{ij}$ = 모든 관측값의 합계
    * 라고 하면... 시그마의 성질에 의해
    * SST = $\sum_{i=1}^k$ $\sum_{j=1}^{n_{i}} y_{ij}^2$ - $\frac{T^2}{n}$, n = $\sum_{i=1}^k n_{i}$
        * $\sum_{i=1}^k$ $\sum_{j=1}^{n_{i}} (y_{ij}-\bar{y})^2$ = $\sum_{i=1}^k$ $\sum_{j=1}^{n_{i}} y_{ij}^2$ -
        2$\bar{y}$ $\sum_{i=1}^k$ $\sum_{j=1}^{n_{i}} y_{ij}$ + $n$ $\bar{y}^2$
        여기서 $\bar{y}$ 는 $\frac{T}{n}$ 이며, $\sum_{i=1}^k$ $\sum_{j=1}^{n_{i}} y_{ij}$ = $T$ 이므로,
        * $\sum_{i=1}^k$ $\sum_{j=1}^{n_{i}} y_{ij}^2$ - $\frac{T^2}{n}$
    * SStr = $\sum_{i=1}^k \frac{T_{i}^2}{n_{i}}$ - $\frac{T^2}{n}$
    * SSE = SST - SStr = $\sum_{i=1}^k$ $\sum_{j=1}^{n_{i}} y_{ij}^2$ - $\sum_{i=1}^k \frac{T_{i}^2}{n_{i}}$
    
    * 예제 2

## 14.3 일원배치 분산분석모형에서의 추론

* 위에서 제곱합에 대한 내용과 처리제곱에 대한 내용을 학습
* 여기서는 이 값들을 이용한 모형에 대해 학습
* 처리(treatment) k 개를 비교하기 위한 모형으로 F 분포를 활용하기 위해 아래와 같이 가정
    * 비교하려는 처리가 k 개이고, 각 처리에서의 반복 측정을 $n_i$ (i = 1, ..., k) 번 할때
    * i 번째 처리에서 j 번째 관측한 반응값을 $Y_{ij}$ 라고 하면...
    * $Y_{ij} = \mu_i + \epsilon_{ij}$, j=1,...,$n_i$, i=1,...,k
    * $\mu_i$ 는 i 번째 처리 모평균을, 오차항 $\epsilon_{ij}$ 는 모두 서로 독립이며 평균이 0, 분산이 $\sigma^2$ 인 정규분포를 따름

* F 분포
    * k 개의 모집단의 모평균이 차이가 없다는 것을 귀무가설로 설정
        * $H_0$ : $\mu_1$ = $\mu_2$ = ... = $\mu_k$
        * $H_1$ : 모든 $\mu_i$ 가 다 같은것은 아니다
    * 귀무가설이 같다면 각 모집단의 평균값과 전체 평균의 값들이 작아질것이므로 ($\bar{y_i}$ - $\bar{y}$) 값들이 작아지게 되고
    * 이 값의 평균처리제곱(MStr)인 $\sum n_i(\bar{y_i} - \bar{y})^2 / (k-1)$ 도 작아지리라 예상할 수 있음
    * 따라서, 평균처리제곱의 크고작음으로 귀무가설의 기각여부를 판단하는데 그 기준으로 공통분산의 추정치인 평균오차제곱(MSE = $s^2$)을 활용한다.
    * 즉, 평균오차제곱에 대한 평균처리제곱의 비율의 크고작음에 따라 귀무가설의 기각여부를 결정함
    * 이때 이 비율 MStr / MSE 가 검정통계량이 되며, 이 분포는 통계이론에 따라 모평균이 동일하다는 귀무가설하에...
    * 자유도가 $(k-1, n-k)$ 인 F 분포... $F(k-1, n-k)$ 를 따름
    * 정리하면...
      F = $\frac {SStr/(k-1)}{SSE/(n-k)}$ = $\frac {MStr}{MSE}$ $\sim F(k-1,n-k)$
    * 검정통계량의 귀무가설 $H_0$ 의 유의수준 $\alpha$ 의 기각역은  $R : F = \frac {MStr}{MSE} \ge F_{\alpha}(k-1, n-k)$
![F분포1](./images/s14-2.png)
![F분포2](./images/s14-3.png)

## 파이썬 분석 따라하기

#### 예제 1 의 자료를 가지고 파이썬을 이용하여 분산분석을 수행하라

In [6]:
import numpy as np
import pandas as pd

y = np.array(
    [10,15,8,12,15,
     14,18,21,15,
     17,16,14,15,17,15,18,
     12,15,17,15,16,15])

# repeat 는 A, B, C, D 를 각각 5번, 4번, 7번, 6번 반복하는것을 의미
treat = np.repeat(['A','B','C','D'],[5,4,7,6])

data = pd.DataFrame({'y':y,'treat':treat})

data

y treat
0   10     A
1   15     A
2    8     A
3   12     A
4   15     A
5   14     B
6   18     B
7   21     B
8   15     B
9   17     C
10  16     C
11  14     C
12  15     C
13  17     C
14  15     C
15  18     C
16  12     D
17  15     D
18  17     D
19  15     D
20  16     D
21  15     D

In [10]:
from statsmodels.formula.api import ols
import statsmodels.api as sm

# ols(model, data).fit() : 입력한 data 를 model 에 대입
# OLS(Ordinary Least Squares)는 가장 기본적인 결정론적 선형 회귀 방법
# 으로 잔차제곱합(RSS: Residual Sum of Squares)를 최소화하는 가중치 벡터를 행렬 미분으로 구하는 방법임
lmFit = ols('y ~ treat', data = data).fit()

# sm.stats.anova_lm(fitted model) : fitted model 에 대한 분산분석 수행
## model : 분산분석을 수행하려 하는 식으로서 '반응변수~설명변수(처리)' 형식으로 입력
## data : 반응변수와 설명변수가 들어있는 데이터 프레임
## fitted model : data 가 적용된 model
sm.stats.anova_lm(lmFit)

df  sum_sq    mean_sq         F    PR(>F)
treat      3.0    68.0  22.666667  4.340426  0.018136
Residual  18.0    94.0   5.222222       NaN       NaN

위 결과에서 treat 와 Residual 은 분산분석에서 변동의 요인들을 나타낸 것으로 각 변동별 자유도(df), 제곱합(sum_sq), 평균제곱(mean_sq), F값과 유의확률(PR(>F)), 즉 P-Value 를 보여준다.
위에서는 P-Value 가 0.01814 이므로 처리들의 효과는 유의수준 5%나 2.5% 에서도 서로 같지 않다고 결론을 내릴 수 있다.